### PR TITLE
SONARPY-1128 Rule S4423: Weak SSL/TLS protocols should not be used

### PR DESCRIPTION
**Pull request by rudy-regazzoni-sonarsource**

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java [hunk 1]
```diff
@@ -52,9 +59,84 @@ public void initialize(Context context) {
...
+        return argTrace.filter(trace -> trace.hasExpression(isString(SENSITIVE_TLS_SECURITY_POLICY))).isPresent();
+      };
+    }
+
+    private static Predicate<Expression> isSensitiveDictionaryTls(SubscriptionContext ctx) {
+      return expression -> Optional.of(expression)
+          .filter(isDictionaryWithKeyValue(ctx, TLS_SECURITY_POLICY, SENSITIVE_TLS_SECURITY_POLICY))
+          .isPresent();
+    }
+  }
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java:141]: IMO separating the logic of pure Python and CDK-related logic by moving this inner class to `org.sonar.python.checks.cdk.WeakSSLProtocolCheckPart` would make it more readable and maintainable

File: python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java [hunk 2]
```diff
@@ -171,4 +176,42 @@ protected static Predicate<Expression> isFqn(String fqnValue) {
...
+    return keyValuePair -> Optional.of(keyValuePair)
+      .map(KeyValuePair::key)
+      .filter(key -> key.is(Tree.Kind.STRING_LITERAL)).map(StringLiteral.class::cast)
+      .filter(string -> string.trimmedQuotesValue().equals(keyName))
+      .isPresent();
+  }
+
+  private static Predicate<KeyValuePair> isKeyValueString(SubscriptionContext ctx, String str) {
+    return keyValuePair -> ArgumentTrace.build(ctx, keyValuePair.value()).hasExpression(isString(str));
+  }
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java:215]: You should have a look at `org.sonar.python.checks.cdk.ClearTextProtocolsCheckPart.ResolvedKeyValuePair` and its use. Maybe you can make use of some logic to reduce duplications.

File: python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java [hunk 3]
```diff
@@ -52,9 +59,84 @@ public void initialize(Context context) {
...
+    // OpenSearch & ElasticSearch
+    private static final String OPENSEARCH_FQN = "aws_cdk.aws_opensearchservice";
+    private static final String ELASTICSEARCH_FQN = "aws_cdk.aws_elasticsearch";
+    private static final String TLS_SECURITY_POLICY = "tls_security_policy";
+    private static final String SENSITIVE_TLS_SECURITY_POLICY = "Policy-Min-TLS-1-0-2019-07";
+
+    @Override
+    protected void registerFqnConsumer() {
+      // Api gateway
+      checkFqn(APIGATEWAY_FQN + ".DomainName", checkDomainName(isFqn(APIGATEWAY_FQN + ".SecurityPolicy.TLS_1_0")));
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java:87]: Maybe you can move the dot to the constants.

File: python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java [hunk 4]
```diff
@@ -52,9 +59,84 @@ public void initialize(Context context) {
...
+        }
+
+        Optional<ArgumentTrace> argTrace = getArgument(ctx, (CallExpression) expression, TLS_SECURITY_POLICY);
+        if (argTrace.isEmpty()) {
+          return true;
+        }
+
+        return argTrace.filter(trace -> trace.hasExpression(isString(SENSITIVE_TLS_SECURITY_POLICY))).isPresent();
+      };
+    }
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/WeakSSLProtocolCheck.java:134]: This logic somehow duplicates the mechanics of `org.sonar.python.checks.cdk.DisabledESDomainEncryptionCheck#isSensitiveOptionObj`. Maybe you can build a generic utils method for this problem.

File: python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java [hunk 5]
```diff
@@ -171,4 +176,42 @@ protected static Predicate<Expression> isFqn(String fqnValue) {
       .isPresent();
   }
 
+  protected static Predicate<Expression> isString(String str) {
+    return expression ->  Optional.of(expression)
+      .filter(expr -> expr.is(Tree.Kind.STRING_LITERAL)).map(StringLiteral.class::cast)
+      .filter(stringLiteral -> stringLiteral.trimmedQuotesValue().equals(str))
+      .isPresent();
+  }
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java:184]: Duplicates `org.sonar.python.checks.cdk.ClearTextProtocolsCheckPart#isStringValue`

File: python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java [hunk 6]
```diff
@@ -171,4 +176,42 @@ protected static Predicate<Expression> isFqn(String fqnValue) {
...
+
+  protected static Predicate<Expression> isDictionaryWithKeyValue(SubscriptionContext ctx, String key, String val) {
+    return expression -> expression.is(Tree.Kind.DICTIONARY_LITERAL)
+      && asDictionaryKeyValuePairs(expression)
+      .filter(isKey(key))
```
Comment (nils-werner-sonarsource) [inline python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java:189]: Keys in dictionaries are also expressions. So they could also be a referenced value. It's an edge case, but can be handled similarly to the resolution of the values.

File: python-checks/src/test/java/org/sonar/python/checks/WeakSSLProtocolCheckTest.java [hunk 7]
```diff
@@ -33,4 +33,14 @@ public void test() {
...
   }
+
+  @Test
+  public void test_apigateway() {
+    PythonCheckVerifier.verify("src/test/resources/checks/weakSSLProtocolCheck_apigateway.py", new WeakSSLProtocolCheck());
+  }
+
+  @Test
+  public void test_elasticopensearch() {
+    PythonCheckVerifier.verify("src/test/resources/checks/weakSSLProtocolCheck_ElasticOpenSearch.py", new WeakSSLProtocolCheck());
```
Comment (nils-werner-sonarsource) [inline python-checks/src/test/java/org/sonar/python/checks/WeakSSLProtocolCheckTest.java:44]: The file name format is not consistent. You should find a common format. Please refer to other cases.

**Review by nils-werner-sonarsource — APPROVED** [review]
LGTM! Please see all my comments first before addressing them. I guess you can make use of some logic, predicates, and structures that are already located in other checks. I'm happy to discuss their advantages and disadvantages.

File: python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java (hunk 5 above)
Comment (rudy-regazzoni-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java:184]: In that case I guess the best is to remove this method isString and move the method isStringValue from ClearTextProtocolsCheckPart to AbstractCdkResourceCheck, so that we can use it anywhere, right ?

Comment (nils-werner-sonarsource) [reply python-checks/src/main/java/org/sonar/python/checks/cdk/AbstractCdkResourceCheck.java:184]: Yes
